File: packages/python/src/softschema/core/value_domain.py
"""Portable JSON-compatible values and materialized-value normalization."""

from __future__ import annotations

import json
import math
from dataclasses import dataclass
from typing import Any, TypeAlias

MAX_SAFE_INTEGER = 9_007_199_254_740_991
MIB = 1024 * 1024

JsonValue: TypeAlias = None | bool | int | float | str | list["JsonValue"] | dict[str, "JsonValue"]


@dataclass(frozen=True)
class ValidationLimits:
    """Resource budgets applied at untrusted artifact and schema boundaries."""

    max_resource_bytes: int = 8 * MIB
    max_bundle_bytes: int = 64 * MIB
    max_resources: int = 256
    max_nodes_per_resource: int = 100_000
    max_depth: int = 128
    max_scalar_codepoints: int = MIB

    def __post_init__(self) -> None:
        nonnegative = ("max_resource_bytes", "max_bundle_bytes", "max_scalar_codepoints")
        positive = ("max_resources", "max_nodes_per_resource", "max_depth")
        for name in (*nonnegative, *positive):
            value = getattr(self, name)
            if isinstance(value, bool) or not isinstance(value, int):
                raise TypeError(f"{name} must be an integer")
        for name in nonnegative:
            if getattr(self, name) < 0:
                raise ValueError(f"{name} must be nonnegative")
        for name in positive:
            if getattr(self, name) < 1:
                raise ValueError(f"{name} must be positive")


DEFAULT_VALIDATION_LIMITS = ValidationLimits()


class PortableYamlError(ValueError):
    """Base class for stable portable-data boundary failures."""

    def __init__(
        self,
        message: str,
        *,
        path: str = "",
        line: int | None = None,
        column: int | None = None,
    ) -> None:
        super().__init__(message)
        self.path = path
        self.line = line
        self.column = column


class PortableYamlSyntaxError(PortableYamlError):
    """The input is not a single syntactically valid YAML document."""


class PortableValueError(PortableYamlError):
    """The input is outside the bounded JSON-compatible value domain."""


@dataclass
class _Budget:
    limits: ValidationLimits
    nodes: int = 0

    def node(
        self,
        path: tuple[str | int, ...],
        depth: int,
        event: Any = None,
        line_offset: int = 0,
    ) -> None:
        if depth > self.limits.max_depth:
            raise _value_error("maximum depth exceeded", path, event, line_offset)
        self.nodes += 1
        if self.nodes > self.limits.max_nodes_per_resource:
            raise _value_error("maximum node count exceeded", path, event, line_offset)

    def scalar(
        self,
        value: str,
        path: tuple[str | int, ...],
        event: Any = None,
        line_offset: int = 0,
    ) -> None:
        if len(value) > self.limits.max_scalar_codepoints:
            raise _value_error("maximum scalar size exceeded", path, event, line_offset)
        if any(0xD800 <= ord(character) <= 0xDFFF for character in value):
            raise _value_error(
                "string contains an invalid Unicode scalar", path, event, line_offset
            )


def normalize_portable_value(
    value: Any,
    *,
    limits: ValidationLimits = DEFAULT_VALIDATION_LIMITS,
    encoded_size: int | None = None,
) -> tuple[JsonValue, int]:
    """Normalize a materialized value and return its canonical JSON size."""
    if encoded_size is not None and encoded_size > limits.max_resource_bytes:
        raise PortableValueError("maximum resource size exceeded")
    budget = _Budget(limits)
    normalized = _normalize_materialized(value, (), 1, budget, set())
    canonical_size = _canonical_json_size(normalized)
    charged_size = encoded_size if encoded_size is not None else canonical_size
    if charged_size > limits.max_resource_bytes:
        raise PortableValueError("maximum resource size exceeded")
    return normalized, charged_size


def _normalize_materialized(
    value: Any,
    path: tuple[str | int, ...],
    depth: int,
    budget: _Budget,
    active: set[int],
) -> JsonValue:
    budget.node(path, depth)
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        budget.scalar(value, path)
        return value
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_INTEGER:
            raise _value_error("integer is outside the safe range", path)
        return value
    if isinstance(value, float):
        if not math.isfinite(value):
            raise _value_error("number must be finite", path)
        if value.is_integer():
            if abs(value) > MAX_SAFE_INTEGER:
                raise _value_error("rounded integer is outside the safe range", path)
            return int(value)
        return value
    if not isinstance(value, dict | list):
        raise _value_error("value is not JSON-compatible", path)

    identity = id(value)
    if identity in active:
        raise _value_error("cycles are not portable", path)
    active.add(identity)
    try:
        if isinstance(value, list):
            return [
                _normalize_materialized(item, (*path, index), depth + 1, budget, active)
                for index, item in enumerate(value)
            ]
        result: dict[str, JsonValue] = {}
        for key, item in value.items():
            budget.node(path, depth + 1)
            if not isinstance(key, str):
                raise _value_error("mapping keys must be strings", path)
            budget.scalar(key, path)
            result[key] = _normalize_materialized(item, (*path, key), depth + 1, budget, active)
        return result
    finally:
        active.remove(identity)


def _canonical_json_size(value: JsonValue) -> int:
    if value is None:
        return 4
    if isinstance(value, bool):
        return 4 if value else 5
    if isinstance(value, int):
        return len(str(value))
    if isinstance(value, float):
        return len(repr(value))
    if isinstance(value, str):
        return len(json.dumps(value, ensure_ascii=False).encode("utf-8"))
    if isinstance(value, list):
        return 2 + max(0, len(value) - 1) + sum(_canonical_json_size(item) for item in value)
    keys = sorted(value)
    return (
        2
        + max(0, len(keys) - 1)
        + sum(_canonical_json_size(key) + 1 + _canonical_json_size(value[key]) for key in keys)
    )


def _value_error(
    message: str,
    path: tuple[str | int, ...],
    event: Any = None,
    line_offset: int = 0,
) -> PortableValueError:
    mark = getattr(event, "start_mark", None)
    return PortableValueError(
        message,
        path=json_pointer(path),
        line=mark.line + 1 + line_offset if mark is not None else None,
        column=mark.column + 1 if mark is not None else None,
    )


def json_pointer(path: tuple[str | int, ...]) -> str:
    """Render a tuple path as an RFC 6901 JSON Pointer."""
    return "".join(f"/{str(part).replace('~', '~0').replace('/', '~1')}" for part in path)

File: packages/typescript/src/core/value-domain.ts
/** Portable JSON-compatible values and materialized-value normalization. */

const MIB = 1024 * 1024;

export interface JsonObject {
  [key: string]: JsonValue;
}

export type JsonValue = null | boolean | number | string | JsonValue[] | JsonObject;

/** Resource budgets applied at untrusted artifact and schema boundaries. */
export interface ValidationLimits {
  maxResourceBytes: number;
  maxBundleBytes: number;
  maxResources: number;
  maxNodesPerResource: number;
  maxDepth: number;
  maxScalarCodePoints: number;
}

/** Trusted per-call overrides; omitted fields retain the portable defaults. */
export type ValidationLimitOverrides = Partial<ValidationLimits>;

export const DEFAULT_VALIDATION_LIMITS: Readonly<ValidationLimits> = Object.freeze({
  maxResourceBytes: 8 * MIB,
  maxBundleBytes: 64 * MIB,
  maxResources: 256,
  maxNodesPerResource: 100_000,
  maxDepth: 128,
  maxScalarCodePoints: MIB,
});

/** Base class for stable portable-data boundary failures. */
export class PortableYamlError extends Error {
  readonly path: string;
  readonly line: number | null;
  readonly column: number | null;

  constructor(
    message: string,
    options: { path?: string; line?: number | null; column?: number | null } = {},
  ) {
    super(message);
    this.path = options.path ?? "";
    this.line = options.line ?? null;
    this.column = options.column ?? null;
  }
}

/** The input is not a single syntactically valid YAML document. */
export class PortableYamlSyntaxError extends PortableYamlError {}

/** The input is outside the bounded JSON-compatible value domain. */
export class PortableValueError extends PortableYamlError {}

interface Budget {
  limits: ValidationLimits;
  nodes: number;
}

export interface NormalizedValue {
  value: JsonValue;
  sizeBytes: number;
}

/** Resolve and validate caller-provided limit overrides. */
export function resolveValidationLimits(
  overrides: ValidationLimitOverrides = {},
): ValidationLimits {
  const limits = { ...DEFAULT_VALIDATION_LIMITS, ...overrides };
  const nonnegative = ["maxResourceBytes", "maxBundleBytes", "maxScalarCodePoints"] as const;
  const positive = ["maxResources", "maxNodesPerResource", "maxDepth"] as const;
  for (const name of [...nonnegative, ...positive]) {
    if (!Number.isSafeInteger(limits[name])) throw new TypeError(`${name} must be an integer`);
  }
  for (const name of nonnegative) {
    if (limits[name] < 0) throw new RangeError(`${name} must be nonnegative`);
  }
  for (const name of positive) {
    if (limits[name] < 1) throw new RangeError(`${name} must be positive`);
  }
  return limits;
}

/** Normalize a materialized value and return its compact canonical JSON size. */
export function normalizePortableValue(
  value: unknown,
  validationLimits: ValidationLimitOverrides = {},
  encodedSize?: number,
): NormalizedValue {
  const limits = resolveValidationLimits(validationLimits);
  if (encodedSize !== undefined && encodedSize > limits.maxResourceBytes) {
    throw new PortableValueError("maximum resource size exceeded");
  }
  const budget: Budget = { limits, nodes: 0 };
  const normalized = normalizeMaterialized(value, [], 1, budget, new Set());
  const sizeBytes = encodedSize ?? canonicalPortableJsonSize(normalized);
  if (sizeBytes > limits.maxResourceBytes) {
    throw new PortableValueError("maximum resource size exceeded");
  }
  return { value: normalized, sizeBytes };
}

/** Return the UTF-8 size of compact JSON with recursively sorted object keys. */
export function canonicalPortableJsonSize(value: JsonValue): number {
  if (value === null) return 4;
  if (typeof value === "boolean") return value ? 4 : 5;
  if (typeof value === "number") return canonicalNumberText(value).length;
  if (typeof value === "string") return utf8Size(JSON.stringify(value));
  if (Array.isArray(value)) {
    return (
      2 +
      Math.max(0, value.length - 1) +
      value.reduce<number>((size, item) => size + canonicalPortableJsonSize(item), 0)
    );
  }
  const keys = Object.keys(value).sort();
  return (
    2 +
    Math.max(0, keys.length - 1) +
    keys.reduce<number>(
      (size, key) =>
        size +
        utf8Size(JSON.stringify(key)) +
        1 +
        canonicalPortableJsonSize(value[key] as JsonValue),
      0,
    )
  );
}

function normalizeMaterialized(
  value: unknown,
  path: readonly (string | number)[],
  depth: number,
  budget: Budget,
  active: Set<object>,
): JsonValue {
  countNode(budget, path, depth);
  if (value === null || typeof value === "boolean") return value;
  if (typeof value === "string") {
    if ([...value].length > budget.limits.maxScalarCodePoints) {
      throw new PortableValueError("maximum scalar size exceeded", { path: jsonPointer(path) });
    }
    if (hasUnpairedSurrogate(value)) {
      throw new PortableValueError("string contains an invalid Unicode scalar", {
        path: jsonPointer(path),
      });
    }
    return value;
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new PortableValueError("number must be finite", { path: jsonPointer(path) });
    }
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) {
      throw new PortableValueError("integer is outside the safe range", {
        path: jsonPointer(path),
      });
    }
    return Object.is(value, -0) ? 0 : value;
  }
  if (typeof value !== "object" || value === undefined) {
    throw new PortableValueError("value is not JSON-compatible", { path: jsonPointer(path) });
  }
  if (active.has(value)) {
    throw new PortableValueError("cycles are not portable", { path: jsonPointer(path) });
  }
  active.add(value);
  try {
    if (Array.isArray(value)) {
      return value.map((item, index) =>
        normalizeMaterialized(item, [...path, index], depth + 1, budget, active),
      );
    }
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) {
      throw new PortableValueError("value is not JSON-compatible", { path: jsonPointer(path) });
    }
    const result: Record<string, JsonValue> = {};
    // Hidden and symbol metadata is outside the materialized JSON object surface.
    // Enumerable accessors remain forbidden so normalization never invokes user code.
    for (const key of Object.keys(value)) {
      const descriptor = Object.getOwnPropertyDescriptor(value, key);
      if (descriptor === undefined || !("value" in descriptor)) {
        throw new PortableValueError("value is not JSON-compatible", {
          path: jsonPointer([...path, key]),
        });
      }
      countNode(budget, path, depth + 1);
      if ([...key].length > budget.limits.maxScalarCodePoints) {
        throw new PortableValueError("maximum scalar size exceeded", { path: jsonPointer(path) });
      }
      if (hasUnpairedSurrogate(key)) {
        throw new PortableValueError("string contains an invalid Unicode scalar", {
          path: jsonPointer(path),
        });
      }
      Object.defineProperty(result, key, {
        configurable: true,
        enumerable: true,
        value: normalizeMaterialized(descriptor.value, [...path, key], depth + 1, budget, active),
        writable: true,
      });
    }
    return result;
  } finally {
    active.delete(value);
  }
}

function canonicalNumberText(value: number): string {
  const absolute = Math.abs(value);
  if (absolute > 0 && absolute < 1e-4 && !Number.isInteger(value)) {
    const [mantissa, rawExponent] = value.toExponential().split("e") as [string, string];
    const negative = rawExponent.startsWith("-");
    const digits = rawExponent.replace(/^[-+]/, "").padStart(2, "0");
    return `${mantissa}e${negative ? "-" : "+"}${digits}`;
  }
  return String(value);
}

function hasUnpairedSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) return true;
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

function countNode(budget: Budget, path: readonly (string | number)[], depth: number): void {
  if (depth > budget.limits.maxDepth) {
    throw new PortableValueError("maximum depth exceeded", { path: jsonPointer(path) });
  }
  budget.nodes += 1;
  if (budget.nodes > budget.limits.maxNodesPerResource) {
    throw new PortableValueError("maximum node count exceeded", { path: jsonPointer(path) });
  }
}

function jsonPointer(path: readonly (string | number)[]): string {
  return path.map((part) => `/${String(part).replace(/~/g, "~0").replace(/\//g, "~1")}`).join("");
}

function utf8Size(value: string): number {
  return new TextEncoder().encode(value).byteLength;
}
